util/chunk/iterator.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package chunk
+
+// Iterator is used to iterate a number of rows.
+//
+// for row := it.Begin(); row != it.End(); row = it.Next() {
+//     ...
+// }
+type Iterator interface {
+	// Begin resets the cursor of the iterator and returns the first Row.
+	Begin() Row
+
+	// Next returns the next Row.
+	Next() Row
+
+	// End returns the invalid end Row.
+	End() Row
+}
+
+// NewSliceIterator returns a Iterator for Row slice.
+func NewSliceIterator(rows []Row) Iterator {
+	return &sliceIterator{rows: rows}
+}
+
+type sliceIterator struct {
+	rows   []Row
+	cursor int
+}
+
+func (it *sliceIterator) Begin() Row {
+	if len(it.rows) == 0 {
+		return Row{}
+	}
+	it.cursor = 1
```

Comment:
let `it.cursor` point at the current row may be more readable.

Reply:
After `Begin` is called, the current row is the row at index 1.

Reply:
It is equivalent to
```go
row := it.rows[0]
it.cursor = 1
return row
```

Reply:
I mean we can let `e.cursor` point to the row idx of the returned `row`.

Reply:
```go
func (it *listIterator) Next() Row {
	if it.chkCursor == it.li.NumChunks() {
		return Row{}
	}
	chk := it.li.GetChunk(it.chkCursor)
	if it.rowCursor + 1 == chk.NumRows() {
		it.chkCursor++
		it.rowCursor = 0
		if it.chkCursor == it.li.NumChunks() {
			return Row{}
		}
	} else {
		it.rowCursor++
	}
	if it.chkCursor == it.li.NumChunks() {
		return Row{}
	}
	chk = it.li.GetChunk(it.chkCursor)
	return chk.GetRow(it.rowCursor)
}
```

Here is the code of `listIterator` Next when cursor points to the returned Row, I think it's harder to read than the current implementation.

```go
func (it *listIterator) Next() Row {
	if it.chkCursor == it.li.NumChunks() {
		return Row{}
	}
	chk := it.li.GetChunk(it.chkCursor)
	row := chk.GetRow(it.rowCursor)
	it.rowCursor++
	if it.rowCursor == chk.NumRows() {
		it.rowCursor = 0
		it.chkCursor++
	}
	return row
}
```

Reply:
ok